File: apps/api-go/middleware/audit.go
// Package middleware holds cross-cutting Encore middleware, mirroring
// apps/api/middleware/. It is not itself an Encore service — global
// middleware just needs to live somewhere in the app for Encore's static
// analyzer to find the //encore:middleware directive.
package middleware

import (
	"encoding/json"
	"regexp"
	"strings"

	"encore.app/authn"
	"encore.app/internal/auditctx"
	"encore.app/internal/entities"
	"encore.app/internal/logger"
	"encore.app/internal/topics"
	"encore.dev/beta/auth"
	"encore.dev/middleware"
)

// auditRoutes maps a route pattern (":param" segments are wildcards) to the
// resource name written to audit_logs.resource. Mirrors apps/api/middleware/
// audit.ts's AUDIT_MAP, adapted to this app's actual Go routes (PATCH
// instead of PUT for bulk updates, no path params on most bulk endpoints).
var auditRoutes = map[string]string{
	"/units":                         "units",
	"/classes":                       "classes",
	"/buildings":                     "buildings",
	"/rooms":                         "rooms",
	"/material-types":                "material_types",
	"/material-stocks":               "material_stocks",
	"/material-assets":               "material_assets",
	"/students":                      "students",
	"/roles":                         "roles",
	"/permissions":                   "permissions",
	"/users":                         "users",
	"/roles/:roleId/permissions":     "role_permissions",
	"/users/:userId/roles":           "user_roles",
	"/export-templates":              "export_templates",
	"/export-templates/:id":          "export_templates",
	"/transfer-requests":             "transfer_requests",
	"/transfer-requests/:id/approve": "transfer_requests",
	"/transfer-requests/:id/reject":  "transfer_requests",
	"/transfer-requests/:id/cancel":  "transfer_requests",
}

// routeActionOverrides pins the audit action for routes where the generic
// method-based rule in auditAction (POST=create, DELETE=delete, else
// update) would be wrong — mirrors apps/api/middleware/audit.ts's AUDIT_MAP,
// whose entries for /transfer-requests/:id/approve and .../reject use the
// dedicated approve/reject actions, and .../cancel logs as an update since
// it doesn't destroy the request row.
var routeActionOverrides = map[string]entities.AuditAction{
	"/transfer-requests/:id/approve": entities.AuditActionApprove,
	"/transfer-requests/:id/reject":  entities.AuditActionReject,
	"/transfer-requests/:id/cancel":  entities.AuditActionUpdate,
}

// joinTableResources are assign/remove endpoints where every mutating
// method (POST assign, DELETE remove) is logged as an "update" rather than
// distinguishing create/delete, since neither call destroys the underlying
// role/user/permission row — only the association.
var joinTableResources = map[string]bool{
	"role_permissions": true,
	"user_roles":       true,
}

var paramSegment = regexp.MustCompile(`^:\w+$`)

func matchAuditRoute(path string) (resource, pattern string, ok bool) {
	if r, ok := auditRoutes[path]; ok {
		return r, path, true
	}

	reqSegs := strings.Split(strings.Trim(path, "/"), "/")
	for pat, resource := range auditRoutes {
		patSegs := strings.Split(strings.Trim(pat, "/"), "/")
		if len(patSegs) != len(reqSegs) {
			continue
		}
		match := true
		for i, seg := range patSegs {
			if paramSegment.MatchString(seg) {
				continue
			}
			if seg != reqSegs[i] {
				match = false
				break
			}
		}
		if match {
			return resource, pat, true
		}
	}

	return "", "", false
}

func auditAction(method, resource, pattern string) entities.AuditAction {
	if override, ok := routeActionOverrides[pattern]; ok {
		return override
	}
	if joinTableResources[resource] {
		return entities.AuditActionUpdate
	}
	switch method {
	case "POST":
		return entities.AuditActionCreate
	case "DELETE":
		return entities.AuditActionDelete
	default:
		return entities.AuditActionUpdate
	}
}

// sensitiveKeys are stripped from previousValue/newValue before they're
// persisted, regardless of what a handler passed to auditctx.SetContext —
// this is the one place every call site funnels through, matching
// apps/api/middleware/audit.ts's sanitizeAuditValue.
var sensitiveKeys = map[string]bool{"password": true}

func sanitizeAuditValue(v any) any {
	switch val := v.(type) {
	case map[string]any:
		out := make(map[string]any, len(val))
		for k, vv := range val {
			if sensitiveKeys[k] {
				continue
			}
			out[k] = sanitizeAuditValue(vv)
		}
		return out
	case []any:
		out := make([]any, len(val))
		for i, vv := range val {
			out[i] = sanitizeAuditValue(vv)
		}
		return out
	default:
		return v
	}
}

// jsonOrDefault marshals v, falling back to def when v is nil/empty so an
// unset audit field still round-trips as valid JSON, matching how
// apps/api/schema/audit-logs.ts defaults resourceIds/previousValue/newValue.
func jsonOrDefault(v any, def string) entities.JSONText {
	if v == nil {
		return entities.JSONText(def)
	}
	b, err := json.Marshal(v)
	if err != nil || string(b) == "null" {
		return entities.JSONText(def)
	}
	return entities.JSONText(b)
}

//encore:middleware global target=tag:audited
func Audit(req middleware.Request, next middleware.Next) middleware.Response {
	data := req.Data()
	resource, pattern, ok := matchAuditRoute(data.Path)
	if !ok {
		return next(req)
	}

	ctx, auditData := auditctx.WithData(req.Context())
	resp := next(req.WithContext(ctx))

	if resp.Err != nil || !auditData.Set {
		return resp
	}

	var actorUserID *int64
	if authData, ok := auth.Data().(*authn.AuthData); ok && authData != nil {
		id := authData.UserID
		actorUserID = &id
	}

	statusCode := resp.HTTPStatus
	if statusCode == 0 {
		statusCode = 200
	}

	_, err := topics.AuditLogTopic.Publish(req.Context(), &topics.AuditLogEvent{
		ActorUserID:   actorUserID,
		Resource:      resource,
		Action:        string(auditAction(data.Method, resource, pattern)),
		ResourceIds:   jsonOrDefault(auditData.ResourceIds, "[]"),
		Method:        data.Method,
		Path:          data.Path,
		StatusCode:    &statusCode,
		PreviousValue: jsonOrDefault(sanitizeAuditValue(auditData.PreviousValue), "{}"),
		NewValue:      jsonOrDefault(sanitizeAuditValue(auditData.NewValue), "{}"),
	})
	if err != nil {
		logger.ErrorContext(req.Context(), "audit middleware: failed to publish audit log event", "err", err)
	}

	return resp
}

File: apps/api/middleware/audit.ts
import { APICallMeta, currentRequest } from 'encore.dev'
import { middleware } from 'encore.dev/api'
import log from 'encore.dev/log'
import { getAuthData } from '~encore/auth'
import { publishAuditEvent } from '../audit-logs/publish'
import { AuditAction } from '../schema/audit-logs'

interface AuditRouteConfig {
	resource: string
	action: AuditAction
}

// Which mutating routes get an audit_logs row, and how to label them.
// Add an entry here (and call setAuditContext from the matching handler)
// whenever a new resource needs to be audited.
const AUDIT_MAP: Record<string, AuditRouteConfig> = {
	'POST:/students': { resource: 'students', action: 'create' },
	'PATCH:/students': { resource: 'students', action: 'update' },
	'DELETE:/students': { resource: 'students', action: 'delete' },

	'POST:/material-assets': { resource: 'material_assets', action: 'create' },
	'PATCH:/material-assets': { resource: 'material_assets', action: 'update' },
	'DELETE:/material-assets': {
		resource: 'material_assets',
		action: 'delete'
	},

	'POST:/material-types': { resource: 'material_types', action: 'create' },
	'PATCH:/material-types': { resource: 'material_types', action: 'update' },
	'DELETE:/material-types': { resource: 'material_types', action: 'delete' },

	'POST:/material-stocks': { resource: 'material_stocks', action: 'create' },
	'PATCH:/material-stocks': { resource: 'material_stocks', action: 'update' },
	'DELETE:/material-stocks': {
		resource: 'material_stocks',
		action: 'delete'
	},

	'POST:/buildings': { resource: 'buildings', action: 'create' },
	'PATCH:/buildings': { resource: 'buildings', action: 'update' },
	'DELETE:/buildings': { resource: 'buildings', action: 'delete' },

	'POST:/rooms': { resource: 'rooms', action: 'create' },
	'PATCH:/rooms': { resource: 'rooms', action: 'update' },
	'DELETE:/rooms': { resource: 'rooms', action: 'delete' },

	'POST:/units': { resource: 'units', action: 'create' },
	'PATCH:/units': { resource: 'units', action: 'update' },
	'DELETE:/units': { resource: 'units', action: 'delete' },

	'POST:/roles': { resource: 'roles', action: 'create' },
	'PUT:/roles/:id': { resource: 'roles', action: 'update' },
	'DELETE:/roles': { resource: 'roles', action: 'delete' },

	'POST:/permissions': { resource: 'permissions', action: 'create' },

	'POST:/users': { resource: 'users', action: 'create' },
	'PUT:/users': { resource: 'users', action: 'update' },
	'DELETE:/users': { resource: 'users', action: 'delete' },

	'POST:/user-roles/assign': { resource: 'user_roles', action: 'update' },

	'POST:/export-templates': {
		resource: 'export_templates',
		action: 'create'
	},
	'DELETE:/export-templates/:id': {
		resource: 'export_templates',
		action: 'delete'
	},

	'POST:/transfer-requests': {
		resource: 'transfer_requests',
		action: 'create'
	},
	'POST:/transfer-requests/:id/approve': {
		resource: 'transfer_requests',
		action: 'approve'
	},
	'POST:/transfer-requests/:id/reject': {
		resource: 'transfer_requests',
		action: 'reject'
	},
	'POST:/transfer-requests/:id/cancel': {
		resource: 'transfer_requests',
		action: 'update'
	}
}

function matchAuditRoute(
	method: string,
	path: string
): AuditRouteConfig | undefined {
	const key = `${method}:${path}`
	if (AUDIT_MAP[key]) {
		return AUDIT_MAP[key]
	}

	for (const [pattern, config] of Object.entries(AUDIT_MAP)) {
		// Route paths can contain params (":id"), so only split on the first
		// colon - a naive pattern.split(':') truncates "/transfer-requests/:id/approve"
		// down to just "/transfer-requests", silently breaking the regex match below.
		const sepIndex = pattern.indexOf(':')
		const patternMethod = pattern.slice(0, sepIndex)
		const patternPath = pattern.slice(sepIndex + 1)
		if (method !== patternMethod) continue

		const regexPattern = patternPath
			.replace(/:\w+/g, '[^/]+')
			.replace(/\*/g, '.*')
		const regex = new RegExp(`^${regexPattern}$`)
		if (regex.test(path)) {
			return config
		}
	}

	return undefined
}

interface AuditContext {
	resourceIds: Array<number | string>
	previousValue?: unknown
	newValue?: unknown
}

// Handlers pass raw domain objects (often with nested Drizzle relations,
// e.g. transfer_requests.approver: UserDB) into setAuditContext. Rather than
// trust every call site to pre-sanitize, strip sensitive keys here — the one
// place all of them funnel through before the value is persisted.
const SENSITIVE_KEYS = new Set(['password'])

function sanitizeAuditValue(value: unknown, seen = new WeakSet()): unknown {
	if (Array.isArray(value)) {
		return value.map((v) => sanitizeAuditValue(v, seen))
	}
	if (value !== null && typeof value === 'object') {
		if (seen.has(value)) return undefined
		seen.add(value)
		const out: Record<string, unknown> = {}
		for (const [key, val] of Object.entries(value)) {
			if (SENSITIVE_KEYS.has(key)) continue
			out[key] = sanitizeAuditValue(val, seen)
		}
		return out
	}
	return value
}

// Extend the middleware data type
declare module 'encore.dev/api' {
	interface MiddlewareData {
		audit?: AuditContext
	}
}

/**
 * Handlers for audited routes call this before returning, so the audit
 * middleware (which cannot see the parsed request body of typed endpoints)
 * has something to log besides the generic envelope. `req.data` set here is
 * the same object exposed to the handler as `currentRequest().middlewareData`,
 * so the write is visible back in the middleware after `next()` resolves.
 */
export function setAuditContext(context: AuditContext) {
	const callMeta = currentRequest() as APICallMeta
	if (callMeta.middlewareData) {
		callMeta.middlewareData.audit = context
	}
}

export const auditMiddleware = middleware(
	{ target: { auth: true } },
	async (req, next) => {
		const meta = req.requestMeta as APICallMeta | undefined
		const config = meta && matchAuditRoute(meta.method, meta.path)

		const resp = await next(req)

		if (!config || !meta) {
			return resp
		}

		const audit = req.data.audit as AuditContext | undefined

		// Handler never called setAuditContext (e.g. bulk op touched nothing) —
		// nothing to write, so skip publishing entirely.
		if (!audit) {
			return resp
		}

		const authData = getAuthData()
		const actorUserId = authData?.userID
			? Number(authData.userID)
			: undefined

		// HandlerResponse.status is write-only (for overriding the status code),
		// so there's nothing to read here. We only reach this point when next()
		// resolved without throwing, i.e. a successful request.
		//
		// The actual DB write happens out-of-band in the audit-logs service's
		// subscription (audit-logs/subscription.ts), which processes this topic
		// on its own goroutine/process rather than inline in the request path.
		publishAuditEvent({
			actorUserId,
			resource: config.resource,
			action: config.action,
			resourceIds: audit.resourceIds ?? [],
			method: meta.method,
			path: meta.path,
			statusCode: 200,
			previousValue: sanitizeAuditValue(audit.previousValue),
			newValue: sanitizeAuditValue(audit.newValue)
		}).catch((err) => {
			log.error('auditMiddleware: failed to publish audit log event', {
				err
			})
		})

		return resp
	}
)
